### DISCUSSION TITLE
Sampling from disparate ROIs?

### DISCUSSION THREAD

**Discussion by kvenkman** [torchgeo/torchgeo#2201] [Q&A] [answered]
Hi, is there a recommended way to setup samplers/datamodules so that disparate ROIs can be sampled from a large image? For example, in this image: 

![image](https://github.com/user-attachments/assets/15d23246-a80d-4b00-838d-6ac4b0b73c79)

is it possible to sample only from cells A-H for training, and from cells G&I for validation? Assuming that the bounds for each cell are known in advance.

**Comment by adamjstewart** [comment, marked as answer]
Multiple ways actually! 

If you don't care about exactly which cells get assigned where, the easiest solution is [torchgeo.datasets.random_grid_cell_assignment](https://torchgeo.readthedocs.io/en/stable/api/datasets.html#torchgeo.datasets.random_grid_cell_assignment). Note that this acts on a single tile at a time, not on the entire area of the dataset. We should probably think about adding something that does the latter (@pmandiola any interest?)

If you do care, and G&I must be in validation, but you do know the bounds for each cell in advance, then you can use [torchgeo.datasets.roi_split](https://torchgeo.readthedocs.io/en/stable/api/datasets.html#torchgeo.datasets.roi_split) instead. You can pass in ROIs for A-F, G, H, and I. Then you can compute the union of A-F and H, and the union of G and I, and you'll end up with exactly what you're looking for.

Hope this helps! And feel free to contribute new splitters if these don't get you what you want. Or better visualizations for the docs. We really need better visualizations for these docs...

**Comment by kvenkman** [comment]
Oh, perfect! It looks like `roi_split` is what I need. I knew about the union operations on datasets, but providing a sequence of ROIs is exactly what I was missing. I'll try this out and let you know if I run into any issues.

I was completely missing the `Utilities` section of that page (it's so far down!). I'd be happy to contribute to visualizations for the docs - should that be something inline, or in the form of a reference Jupyter notebook?

**Reply by adamjstewart** [reply]
I think both would be fantastic. We could put individual images in the API docs and an entire notebook on dataset splitting in our tutorials. You can store the images in the `images/` directory.
